Add more code splitters (go, rst, js, java, cpp, scala, ruby, php, swift, rust)

ByronHsu opened the pull request:
As the title says, I added more code splitters.
The implementation is trivial, so i don't add separate tests for each splitter. 
Let me know if any concerns.

Fixes # (issue)
https://github.com/hwchase17/langchain/issues/5170

## Who can review?

Community members can review the PR once tests pass. Tag maintainers/contributors who might be interested:
@eyurtsev @hwchase17

Commit:
Add more code splitters (go, rst, js, java, cpp, scala, ruby, php, swift, rst)

Signed-off-by: byhsu <byhsu@linkedin.com>

Commit:
format

Signed-off-by: byhsu <byhsu@linkedin.com>

dev2049 commented:
could we add some unit tests for all these :) may be easier to have sep pr for each, but will let you decide

ByronHsu commented:
@dev2049 ok i will add tests and split the pr :)

Commit:
add tests

Signed-off-by: byhsu <byhsu@linkedin.com>

ByronHsu commented:
Added tests! plz take a look again. Thanks!

btw, may i ask why i run into mypy error locally, but not on github action?

```
~/learn-repo/langchain more-code-splitter !2 ❯ make lint                                                                                              4s  langchain byhsu@byhsu-ld1 22:50:19
poetry run mypy .
langchain/evaluation/loading.py:5: error: Incompatible import of "load_dataset" (imported name has type "Callable[[str, Optional[str], Optional[str], Union[str, Sequence[str], Mapping[str, Union[str, Sequence[str]]], None], Union[str, Split, None], Optional[str], Optional[Features], Optional[DownloadConfig], Optional[GenerateMode], bool, Optional[bool], bool, Union[str, Version, None], Union[bool, str, None], Union[str, TaskTemplate, None], bool, Any, KwArg(Any)], Union[DatasetDict, Dataset, IterableDatasetDict, IterableDataset]]", local name has type "Callable[[str], List[Dict[Any, Any]]]")  [assignment]
langchain/evaluation/loading.py:8: error: No overload variant of "__getitem__" of "list" matches argument type "str"  [call-overload]
langchain/evaluation/loading.py:8: note: Possible overload variants:
langchain/evaluation/loading.py:8: note:     def __getitem__(self, SupportsIndex, /) -> Dict[Any, Any]
langchain/evaluation/loading.py:8: note:     def __getitem__(self, slice, /) -> List[Dict[Any, Any]]
langchain/document_loaders/hugging_face_dataset.py:81: error: Item "Dataset" of "Union[DatasetDict, Dataset, IterableDatasetDict, IterableDataset]" has no attribute "keys"  [union-attr]
langchain/document_loaders/hugging_face_dataset.py:81: error: Item "IterableDataset" of "Union[DatasetDict, Dataset, IterableDatasetDict, IterableDataset]" has no attribute "keys"  [union-attr]
langchain/document_loaders/hugging_face_dataset.py:82: error: Value of type "Union[DatasetDict, Dataset, IterableDatasetDict, IterableDataset]" is not indexable  [index]
Found 5 errors in 2 files (checked 1024 source files)
make: *** [lint] Error 1
```

eyurtsev commented on langchain/text_splitter.py:
```diff
@@ -439,3 +439,292 @@ def __init__(self, **kwargs: Any):
         super().__init__(separators=separators, **kwargs)
+
+
+class GolangCodeTextSplitter(RecursiveCharacterTextSplitter):
```
@ByronHsu @dev2049 

Instead of creating new classes for each language, could we parameterize by language? 

It'll save at least for lines of code per new language that's introduced, and also make it easier to see what languages are available.

```python

from typing import Union, Literal

Language = Union[Literal['go'], Literal['python'], Literal[...]]

class CodeTextSplitter(RecursiveCharacterTextSplitter):
    def __init__(self, language: Language, **kwargs: Any):
        separators = get_separators_for_language(language)
        super().__init__(separators=separators, **kwargs)
```

ByronHsu replied:
Good idea! @dev2049 wdyt?

dev2049 replied:
im a fan! do we prefer typing language being a `Literal['python', 'go', ...]` or just making a string enum
```python
class Language(str, Enum):
  PYTHON = "python
  GO = "go"
  ...
```

ByronHsu replied:
@dev2049 @eyurtsev Updated per our discussion!

Commit:
use enum

Signed-off-by: byhsu <byhsu@linkedin.com>

Commit:
lint

Signed-off-by: byhsu <byhsu@linkedin.com>

ByronHsu commented:
@eyurtsev Can you review? Thanks!

eyurtsev commented on langchain/text_splitter.py:
```diff
@@ -439,3 +440,261 @@ def __init__(self, **kwargs: Any):
         super().__init__(separators=separators, **kwargs)
+
+
+class Language(str, Enum):
```
This looks good to me. 

Have you considered using a union of literals instead of an enum?

```
Language = Union[Literal['go'], Literal['js'], ...]
```

Gives the same static typing guarantees and makes it such that a user does not have to import an extra enum.

I think the only downside is that it's harder to list all the options in the set.

ByronHsu replied:
Hmm I would think prompting available languages is more important to users

corleyma replied:
I believe VSCode/PyLance would still prompt appropriately for a union of literals, but the enum is more foolproof.

eyurtsev reviewed (commented):
Left some minor suggestion with enum. Let's update the exception to a ValueError rather than throwing a generic exception.

eyurtsev commented:
@ByronHsu i can take a look at mypy error on monday. i assume you checked from master?

corleyma commented on langchain/text_splitter.py:
```diff
@@ -439,3 +440,261 @@ def __init__(self, **kwargs: Any):
+class CodeTextSplitter(RecursiveCharacterTextSplitter):
+    def __init__(self, language: Language, **kwargs: Any):
+        """
+        A generic code text splitter supporting many programming languages.
```
why not include markdown and latex?  They are already here, could still expose those classes for backwards compatibility but would also make sense to consolidate them here.

ByronHsu replied:
Sounds great! I have added them

Commit:
Add latex and markdown, and use valueerror

Signed-off-by: byhsu <byhsu@linkedin.com>

Commit:
lint

Signed-off-by: byhsu <byhsu@linkedin.com>

Commit:
lint

Signed-off-by: byhsu <byhsu@linkedin.com>

Commit:
resolve conflict

Signed-off-by: byhsu <byhsu@linkedin.com>

Commit:
Improve test

Signed-off-by: byhsu <byhsu@linkedin.com>

Commit:
add nb

Signed-off-by: byhsu <byhsu@linkedin.com>

Commit:
add nb

Signed-off-by: byhsu <byhsu@linkedin.com>

ByronHsu commented:
@dev2049 i've added notebook examples and improved the tests. Could you plz review again? Thanks!
